### PR TITLE
feat: stage status cache

### PR DESCRIPTION
**Pull request by kuisathaverat**
## What does this PR do?

It adds a new step to save/check the last execution status of a stage-based in its name and commit.

## Why is it important?

This allows skipping stages for the same commit in case of retries.

### CODE REVIEW

**Review by cachedout — APPROVED** [review]

File: vars/stageStatusCache.groovy [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/**
+  Execute the body if there is not stage status file for the stage and the commit we are building.
+  User triggered builds will execute all stages always.
+  If the stage success the status is save in a file.
+*/
+def stageStatusCache(Map args, Closure body){
```
Comment (v1v) [inline vars/stageStatusCache.groovy:23]:
```suggestion
def call(Map args, Closure body){
```

File: vars/stageStatusCache.groovy [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  if(!args.containsKey('bucket')){
+    args['bucket'] = 'beats-ci-temp'
+  }
+  if(!args.containsKey('credentialsId')){
+    args['credentialsId'] = 'beats-ci-gcs-plugin-file-credentials'
+  }
+  if(!args.containsKey('sha')){
+    args['sha'] = "${GIT_BASE_COMMIT}"
+  }
+  if(readStageStatus(args) == false || isUserTrigger() || env.BUILD_ID == "1" || env.RUN_ALL == "1"){
```
Comment (v1v) [inline vars/stageStatusCache.groovy:33]: `RUN_ALL` is quite specific, do you think we can use an argument instead? Something like

```suggestion
  if(readStageStatus(args) == false || isUserTrigger() || env.BUILD_ID == "1" || args.get('runAlways', false)){
```

File: vars/stageStatusCache.groovy [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  Read the status file of the stage if it exists.
+*/
+def readStageStatus(Map args){
+  def statusFileName = stageStatusId(args)
+  try {
+    cmd(label: 'Download Stage Status',
+      script: "curl -sSf -O https://storage.googleapis.com/${args.bucket}/ci/cache/${statusFileName}",
+      returnStatus: true)
+  } catch(e) {
+    log(level: 'WARN', text: "There is no cache file for the current stage.")
```
Comment (v1v) [inline vars/stageStatusCache.groovy:63]: I guess, this should never happen as long as the `returnStatus: true` is set

File: vars/stageStatusCache.groovy (hunk 1 above)
Comment (v1v) [inline vars/stageStatusCache.groovy:23]: Missing the list of supported arguments:

- bucket
- credentialsId
- sha
- id   < ---- this should be mandatory

File: vars/stageStatusCache.txt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        }
+      }
+    }
+  }
+}
+```
+
+* *bucket:* bucket name.
+* *credentialsId:* credentials file, with the GCP credentials JSON file.
+* *sha:* Commit SHA used for the stage ID.
```
Comment (v1v) [inline vars/stageStatusCache.txt:33]:
```suggestion
* *id:* Unique stage name. Mandatory
* *bucket:* bucket name. Default 'beats-ci-temp'
* *credentialsId:* credentials file, with the GCP credentials JSON file. Default  'beats-ci-gcs-plugin-file-credentials'
* *sha:* Commit SHA used for the stage ID. Default: env.GIT_BASE_COMMIT
```

File: vars/stageStatusCache.groovy (hunk 1 above)
Comment (v1v) [reply vars/stageStatusCache.groovy:23]: https://github.com/elastic/apm-pipeline-library/pull/1061#discussion_r603894032

File: vars/stageStatusCache.groovy (hunk 3 above)
Comment (kuisathaverat) [reply vars/stageStatusCache.groovy:63]: yep

**Review by v1v — APPROVED** [review]

**Pull request merged**